pkg/liquidity-mining/test/GaugeWorkingBalanceHelper.test.ts

```diff
@@ -0,0 +1,339 @@
+        );
+
+        expect(projected).to.gt(current);
+      });
+    }
+
+    function projectedRatioIsMaxedOut() {
+      it(`projected ratio should be greater than current by the maximum ratio (${MAX_BALANCE_RATIO})`, async () => {
+        const [currentWorkingBalance, projectedWorkingBalance] = await workingBalanceHelper.getWorkingBalances(
+          gauge.address,
+          user.address
+        );
+
+        expect(fromFp(projectedWorkingBalance) / fromFp(currentWorkingBalance)).to.eq(MAX_BALANCE_RATIO);
+      });
+    }
+
+    function projectedBalanceIsIncreased(maxPctIncrease: number) {
+      it(`projected balance should be slightly more than current, by less than ${maxPctIncrease}%`, async () => {
+        const [currentWorkingBalance, projectedWorkingBalance] = await workingBalanceHelper.getWorkingBalances(
+          gauge.address,
+          user.address
+        );
+
+        expect(projectedWorkingBalance).to.be.gt(currentWorkingBalance);
+
+        const ratio = fromFp(projectedWorkingBalance) / fromFp(currentWorkingBalance);
+
+        expect(ratio).to.be.lt(1 + maxPctIncrease / 100);
+      });
+    }
+
+    function currentShouldAlmostEqualProjected() {
+      it('projected balance should be close to or less than current', async () => {
+        const [currentWorkingBalance, projectedWorkingBalance] = await workingBalanceHelper.getWorkingBalances(
+          gauge.address,
+          user.address
+        );
+
+        expect(projectedWorkingBalance).to.be.almostEqual(currentWorkingBalance);
+        expect(projectedWorkingBalance).to.be.gt(0);
+        expect(projectedWorkingBalance).to.be.lte(currentWorkingBalance);
+      });
+    }
+
+    function itUpdatesAfterCheckpointing() {
+      context('when checkpointed', () => {
+        sharedBeforeEach('checkpoint user', async () => {
+          await gauge.connect(user).user_checkpoint(user.address);
+        });
+
+        currentShouldAlmostEqualProjected();
+      });
+    }
+
+    function veBALDecaysOverTime() {
+      it('veBAL decays over time', async () => {
+        const [, projecteBalanceBefore] = await workingBalanceHelper.getWorkingBalances(gauge.address, user.address);
+        const [, projectedRatioBefore] = await workingBalanceHelper.getWorkingBalanceToSupplyRatios(
+          gauge.address,
+          user.address
+        );
+
+        // Checkpoint user, add another contribution to maintain the total supply, and advance time
+        await gauge.connect(user).user_checkpoint(user.address);
+        await createLockForUser(another, bptAmount.mul(100), LOCK_PERIOD);
+        await advanceTime(LOCK_PERIOD / 2);
+
+        const [, projectedBalanceAfter] = await workingBalanceHelper.getWorkingBalances(gauge.address, user.address);
+        const [, projectedRatioAfter] = await workingBalanceHelper.getWorkingBalanceToSupplyRatios(
+          gauge.address,
+          user.address
+        );
+
+        // Projections should be uniformly lower
```

Comment:
What about projected before vs current after? Shouldn't they be equal after the checkpoint?

That could also be a good indication that the test is correct.

Reply:
Added